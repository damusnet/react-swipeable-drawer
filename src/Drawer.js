import { Component } from "react";
import PropTypes from "prop-types";

class Drawer extends Component {
  static propTypes = {
    width: PropTypes.number.isRequired,
    children: PropTypes.func.isRequired,
  };

  state = {
    swiping: false,
    scrolling: false,
    translateX: 0,
    clientX: 0,
    clientY: 0,
  };

  componentDidMount() {
    window.addEventListener("scroll", this.onScroll);
  }

  componentWillUnmount() {
    window.removeEventListener("scroll", this.onScroll);
  }

  onScroll = () => {
    const { translateX } = this.state;
    if (translateX === 0) {
      this.mainContentScroll = window.pageYOffset;
    }
  };

  mainContentScroll = 0;

  toggleDrawer = () => {
    this.setState(
      ({ translateX }) => ({
        translateX: translateX > 50 ? 0 : 100,
      }),
      this.restoreMainContentScrollPosition
    );
  };

  handleTouchStart = event => {
    const { clientX, clientY } = event.targetTouches[0];
    this.setState({ swiping: true, clientX, clientY });
  };

  handleTouchMove = width => event => {
    const {
      clientX: prevClientX,
      clientY: prevClientY,
      scrolling,
    } = this.state;
    const maxWidth = window.innerWidth;
    const { clientX, clientY } = event.targetTouches[0];

    const diffTranslateX = Math.abs(clientX - prevClientX);
    const diffTranslateY = Math.abs(clientY - prevClientY);

    if (scrolling || diffTranslateY > diffTranslateX) {
      this.setState({ scrolling: true });
    } else {
      const translateX = Math.min(
        clientX / (maxWidth * width / 100) * 100,
        100
      );
      this.setState({ translateX });
    }
  };

  handleTouchEnd = () => {
    this.setState(
      ({ translateX }) => ({
        swiping: false,
        scrolling: false,
        translateX: translateX < 50 ? 0 : 100,
      }),
      this.restoreMainContentScrollPosition
    );
  };

  restoreMainContentScrollPosition = () => {
    const { translateX } = this.state;
    if (translateX === 0) {
      window.scrollTo(0, this.mainContentScroll);
    }
  };

  render() {
    const { width, children } = this.props;
    const { swiping, translateX } = this.state;

    return children({
      width,
      swiping,
      translateX,
      mainContentScroll: this.mainContentScroll,
      toggleDrawer: this.toggleDrawer,
      handleTouchStart: this.handleTouchStart,
      handleTouchMove: this.handleTouchMove,
      handleTouchEnd: this.handleTouchEnd,
    });
  }
}

export default Drawer;

export { default as DrawerContainer } from "./DrawerContainer";
export { default as DrawerOverlay } from "./DrawerOverlay";
export { default as DrawerContentContainer } from "./DrawerContentContainer";
export { default as MainContentContainer } from "./MainContentContainer";
