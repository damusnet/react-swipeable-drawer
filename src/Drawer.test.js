import React from "react";
import { shallow, mount } from "enzyme";

import Drawer from "./Drawer";

const noop = () => <div />;

describe("<Drawer />", () => {
  it("toggles open", () => {
    const drawer = shallow(
      <Drawer size={80} position="left" zIndex={1}>
        {noop}
      </Drawer>
    ).instance();

    drawer.toggleDrawer();

    expect(drawer.state.translation).toEqual(100);
  });

  it("toggles close", () => {
    const drawer = shallow(
      <Drawer size={80} position="left" zIndex={1}>
        {noop}
      </Drawer>
    ).instance();

    drawer.toggleDrawer();
    drawer.toggleDrawer();

    expect(drawer.state.translation).toEqual(-10);
  });

  it("swipes horizontally", () => {
    const drawer = shallow(
      <Drawer size={80} position="left" zIndex={1}>
        {noop}
      </Drawer>
    ).instance();

    drawer.handleTouchStart({
      targetTouches: [{ clientX: 100, clientY: 100 }],
    });

    drawer.handleTouchMove(80)({
      targetTouches: [{ clientX: 900, clientY: 100 }],
    });

    expect(drawer.state.translation).toEqual(100);
  });

  it("scrolls vertically", () => {
    const drawer = shallow(
      <Drawer size={80} position="left" zIndex={1}>
        {noop}
      </Drawer>
    ).instance();

    drawer.handleTouchStart({
      targetTouches: [{ clientX: 100, clientY: 100 }],
    });

    drawer.handleTouchMove(80)({
      targetTouches: [{ clientX: 100, clientY: 900 }],
    });

    expect(drawer.state.scrolling).toEqual(true);

    drawer.handleTouchMove(80)({
      targetTouches: [{ clientX: 200, clientY: 900 }],
    });

    expect(drawer.state.translation).toEqual(-10);
  });

  it("closes when swiping a little", () => {
    const drawer = shallow(
      <Drawer size={80} position="left" zIndex={1}>
        {noop}
      </Drawer>
    ).instance();

    drawer.handleTouchStart({
      targetTouches: [{ clientX: 100, clientY: 100 }],
    });

    drawer.handleTouchMove(80)({
      targetTouches: [{ clientX: 200, clientY: 100 }],
    });

    drawer.handleTouchEnd();

    expect(drawer.state.translation).toEqual(-10);
  });

  it("opens when swiping enough from left to right", () => {
    const drawer = shallow(
      <Drawer size={80} position="left" zIndex={1}>
        {noop}
      </Drawer>
    ).instance();

    drawer.handleTouchStart({
      targetTouches: [{ clientX: 100, clientY: 100 }],
    });

    drawer.handleTouchMove(80)({
      targetTouches: [{ clientX: 500, clientY: 100 }],
    });

    drawer.handleTouchEnd();

    expect(drawer.state.translation).toEqual(100);
  });

  it("opens when swiping enough from right to left", () => {
    const drawer = shallow(
      <Drawer size={80} position="right" zIndex={1}>
        {noop}
      </Drawer>
    ).instance();

    drawer.handleTouchStart({
      targetTouches: [{ clientX: 500, clientY: 100 }],
    });

    drawer.handleTouchMove(80)({
      targetTouches: [{ clientX: 100, clientY: 100 }],
    });

    drawer.handleTouchEnd();

    expect(drawer.state.translation).toEqual(100);
  });

  it("resets the window scroll position", () => {
    window.scrollTo = jest.fn();
    window.pageYOffset = 100;

    const drawer = mount(
      <Drawer size={80} position="left" zIndex={1}>
        {noop}
      </Drawer>
    ).instance();

    drawer.toggleDrawer();
    drawer.toggleDrawer();

    expect(drawer.mainContentScroll).toEqual(100);
    expect(window.scrollTo).toBeCalledWith(0, 100);
  });
});
