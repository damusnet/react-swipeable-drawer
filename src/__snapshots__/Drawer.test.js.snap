// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Drawer /> closes when swiping a little 1`] = `
<div>
  <div
    className="DrawerContainer"
  >
    <DrawerOverlay
      handleTouchEnd={[Function]}
      handleTouchMove={[Function]}
      handleTouchStart={[Function]}
      open={false}
      swiping={false}
      toggleDrawer={[Function]}
      translateX={0}
    />
    <DrawerContent
      drawerContent={<div />}
      handleTouchEnd={[Function]}
      handleTouchMove={[Function]}
      handleTouchStart={[Function]}
      swiping={false}
      toggleDrawer={[Function]}
      translateX={0}
      width={80}
    />
  </div>
  <div
    style={Object {}}
    toggleDrawer={[Function]}
  />
</div>
`;

exports[`<Drawer /> opens when swiping enough 1`] = `
<div>
  <div
    className="DrawerContainer"
  >
    <DrawerOverlay
      handleTouchEnd={[Function]}
      handleTouchMove={[Function]}
      handleTouchStart={[Function]}
      open={true}
      swiping={false}
      toggleDrawer={[Function]}
      translateX={100}
    />
    <DrawerContent
      drawerContent={<div />}
      handleTouchEnd={[Function]}
      handleTouchMove={[Function]}
      handleTouchStart={[Function]}
      swiping={false}
      toggleDrawer={[Function]}
      translateX={100}
      width={80}
    />
  </div>
  <div
    style={
      Object {
        "position": "fixed",
        "top": -0,
      }
    }
    toggleDrawer={[Function]}
  />
</div>
`;

exports[`<Drawer /> swipes horizontally 1`] = `
<div>
  <div
    className="DrawerContainer"
  >
    <DrawerOverlay
      handleTouchEnd={[Function]}
      handleTouchMove={[Function]}
      handleTouchStart={[Function]}
      open={true}
      swiping={true}
      toggleDrawer={[Function]}
      translateX={24.4140625}
    />
    <DrawerContent
      drawerContent={<div />}
      handleTouchEnd={[Function]}
      handleTouchMove={[Function]}
      handleTouchStart={[Function]}
      swiping={true}
      toggleDrawer={[Function]}
      translateX={24.4140625}
      width={80}
    />
  </div>
  <div
    style={
      Object {
        "position": "fixed",
        "top": -0,
      }
    }
    toggleDrawer={[Function]}
  />
</div>
`;

exports[`<Drawer /> swipes vertically 1`] = `
<div>
  <div
    className="DrawerContainer"
  >
    <DrawerOverlay
      handleTouchEnd={[Function]}
      handleTouchMove={[Function]}
      handleTouchStart={[Function]}
      open={false}
      swiping={true}
      toggleDrawer={[Function]}
      translateX={0}
    />
    <DrawerContent
      drawerContent={<div />}
      handleTouchEnd={[Function]}
      handleTouchMove={[Function]}
      handleTouchStart={[Function]}
      swiping={true}
      toggleDrawer={[Function]}
      translateX={0}
      width={80}
    />
  </div>
  <div
    style={Object {}}
    toggleDrawer={[Function]}
  />
</div>
`;

exports[`<Drawer /> toggles close 1`] = `
<div>
  <div
    className="DrawerContainer"
  >
    <DrawerOverlay
      handleTouchEnd={[Function]}
      handleTouchMove={[Function]}
      handleTouchStart={[Function]}
      open={false}
      swiping={false}
      toggleDrawer={[Function]}
      translateX={0}
    />
    <DrawerContent
      drawerContent={<div />}
      handleTouchEnd={[Function]}
      handleTouchMove={[Function]}
      handleTouchStart={[Function]}
      swiping={false}
      toggleDrawer={[Function]}
      translateX={0}
      width={80}
    />
  </div>
  <div
    style={Object {}}
    toggleDrawer={[Function]}
  />
</div>
`;

exports[`<Drawer /> toggles open 1`] = `
<div>
  <div
    className="DrawerContainer"
  >
    <DrawerOverlay
      handleTouchEnd={[Function]}
      handleTouchMove={[Function]}
      handleTouchStart={[Function]}
      open={true}
      swiping={false}
      toggleDrawer={[Function]}
      translateX={100}
    />
    <DrawerContent
      drawerContent={<div />}
      handleTouchEnd={[Function]}
      handleTouchMove={[Function]}
      handleTouchStart={[Function]}
      swiping={false}
      toggleDrawer={[Function]}
      translateX={100}
      width={80}
    />
  </div>
  <div
    style={
      Object {
        "position": "fixed",
        "top": -0,
      }
    }
    toggleDrawer={[Function]}
  />
</div>
`;
